runtime/lua/vim/shared.lua

```diff
@@ -1139,4 +1139,61 @@ function vim._defer_require(root, mod)
   })
 end
 
+--- @nodoc
+--- @class vim.context.mods
+--- @field sandbox? boolean
+--- @field noautocmd? boolean
+--- @field hide? boolean
+--- @field horizontal? boolean
+--- @field keepalt? boolean
+--- @field keepjumps? boolean
+--- @field keepmarks? boolean
+--- @field keeppatterns? boolean
+--- @field lockmarks? boolean
+---
+--- @field buf? integer
+--- @field win? integer
+---
+--- @field options? table<string, any>
+
+--- Wrapper over the `vim._with_c` function to restore editor state that is difficult to do from
+--- the C code.
+---
+--- @param context vim.context.mods
+function vim._with(context, f)
+  -- Check buffer exists
+  if context.buf then
+    if not vim.api.nvim_buf_is_valid(context.buf) then
```

Comment:
```suggestion
    if not (type(context.buf) == 'number' and vim.api.nvim_buf_is_valid(context.buf)) then
```

This will give nicer error messages if `context.buf` is not a number.

Same for `context.win` check.

Reply:
This is internal for now so callers are expected to call this with correct argument types. Gating with extra conditions can hide bugs in other code. So check with `vim.validate` or not at all.

Reply:
I guess it makes sense to use `vim.validate` since we're eventually planning to expose this, and vim.validate will be used then anyway.

Reply:
It doesn't look like vim.validate can validate the fields of a table (at least there were no indications of it nor any examples of it). I guess we'll not add anything then.